src/graph.rs

```diff
@@ -0,0 +1,343 @@
+            values: SmallVec::new(),
+        }
+    }
+
+    /// Adds an attribute to this attribute set.  If there was already an attribute with the same
+    /// name, replaces its value and returns `Err`.
+    pub fn add<V: Into<Value>>(&mut self, name: Identifier, value: V) -> Result<(), ()> {
+        match self.values.binary_search_by_key(&name, |(name, _)| *name) {
+            Ok(_) => Err(()),
+            Err(index) => {
+                self.values.insert(index, (name, value.into()));
+                Ok(())
+            }
+        }
+    }
+
+    /// Returns the value of a particular attribute, if it exists.
+    pub fn get(&self, name: Identifier) -> Option<&Value> {
+        self.values
+            .binary_search_by_key(&name, |(name, _)| *name)
+            .ok()
+            .map(|index| &self.values[index].1)
+    }
+
+    /// Displays the contents of this attribute set.
+    pub fn display_with<'a>(&'a self, ctx: &'a Context, graph: &'a Graph) -> impl Display + 'a {
+        struct DisplayAttributes<'a, 'tree>(&'a Attributes, &'a Context, &'a Graph<'tree>);
+
+        impl<'a, 'tree> Display for DisplayAttributes<'a, 'tree> {
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                let attributes = self.0;
+                let ctx = self.1;
+                let graph = self.2;
+                for (name, value) in &attributes.values {
+                    write!(
+                        f,
+                        "  {}: {}\n",
+                        ctx.resolve(*name),
+                        value.display_with(graph),
+                    )?;
+                }
+                Ok(())
+            }
+        }
+
+        DisplayAttributes(self, ctx, graph)
+    }
+}
+
+/// The value of an attribute
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub enum Value {
+    // Scalar
+    Null,
+    Boolean(bool),
+    Integer(u32),
+    String(String),
+    // Compound
+    List(Vec<Value>),
+    Set(BTreeSet<Value>),
+    // References
+    SyntaxNode(SyntaxNodeRef),
+    GraphNode(GraphNodeRef),
+}
+
+impl Value {
+    /// Displays this value.
+    pub fn display_with<'a, 'tree>(&'a self, graph: &'a Graph<'tree>) -> impl Display + 'a {
+        struct DisplayValue<'a, 'tree>(&'a Value, &'a Graph<'tree>);
+
+        impl<'a, 'tree> Display for DisplayValue<'a, 'tree> {
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                let graph = self.1;
+                match self.0 {
+                    Value::Null => write!(f, "#null"),
+                    Value::Boolean(value) => {
+                        if *value {
+                            write!(f, "#true")
+                        } else {
+                            write!(f, "#false")
+                        }
+                    }
+                    Value::Integer(value) => write!(f, "{}", value),
+                    Value::String(value) => write!(f, "{}", value),
```

Comment:
```suggestion
                    Value::String(value) => write!(f, "\"{}\"", value.escape_default()),
```
This makes the output less confusing when the string contains a newline, and also disambiguates e.g. the integer `5` from the string `"5"`.

Reply:
It does: https://github.com/tree-sitter/tree-sitter-graph/blob/5461b8bf988caecde48650153ea652c49ba0faf8/tests/it/graph.rs#L69-L71